server/aesdsocket.c

```diff
@@ -1,82 +1,203 @@
-
-signal(SIGINT, signal_handler);
-signal(SIGTERM, signal_handler);
-/******************************************************************************************** Create the Socket **************************************************************************************/
-syslog(LOG_USER, "Socket Creation");
-sockfd = socket(AF_INET,SOCK_STREAM,0);
-if(sockfd == -1){
-   perror("Socket Not created Properly");
-   exit(-1);
+typedef TAILQ_HEAD(head_s, node) head_t;
+head_t head;
+// Assignment instruction 2b - Append timestamp in 24hours format
+// parameters : 	None
+// Returns    :	None
+// References	:	https://geeksforgeeks.org/strftime-function-in-c/
+void append_time_stamp()
+{
+	time_t t;
+	struct tm *tmp;
+	char MY_TIME[TIMESTAMP_SIZE];
+	time(&t);
+	tmp = localtime(&t);
+	strftime(MY_TIME, sizeof(MY_TIME), "timestamp: %Y-%m-%d %H:%M:%S\r\n", tmp);
+	printf("%s", MY_TIME);
+	lseek(fd, 0, SEEK_END);						// Move the current file to the end
+	int write_bytes_count = write(fd, MY_TIME, strlen(MY_TIME));
+	if (write_bytes_count != strlen(MY_TIME)) {
+		printf("write unsuccessful\n");
+	}
 }
-/******************************************************************************************** Fill in the Data Structure *****************************************************************************/
-struct sockaddr_in server;
-memset((void *)&server, 0, sizeof(server));
-server.sin_family = AF_INET;
-server.sin_addr.s_addr = INADDR_ANY;
-server.sin_port = htons(9000);
-/***************************************************************************************** Bind the client with the server ***************************************************************************/
-int yes = 1;
-if(setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes)) == -1){
-	perror("Unable to Setup reusing ability of the socket port and ip");
-	exit(-1);
+// Remove all the memory to avoid memory leaks from valgrind checks, in this program, that occurs only during SIGINT, SIGTERM
+// queue.h functions handbook line 187 - 192
+void delete_all_the_memory()
+{
+	thread_data_t *datap;
+	while(!TAILQ_EMPTY(&head))
+	{
+		datap = TAILQ_FIRST(&head);
+		TAILQ_REMOVE(&head, datap, entries);
+		free(datap);
+	}
+}
+// Signal handler function to terminate during SIGINT and SIGTERM and add timestamp every 10 seconds
+void signal_handler(int signo)
+{
+	printf("%d signal caught", signo);
+	// Call timestamp every 10 seconds despite other operations happening in parallel, discussed in robert love linux system programming chapter 10, Page 387
+	if (signo == SIGALRM)
+	{
+		printf("SIGALRM ALERT\n");
+		append_time_stamp();
+		alarm(10);
+	}
+	// In case a SIGINT or SIGTERM happens (pressing ctrl + c, graceful cleanup operation)
+	// Perform program exit - As per assignment instructions 1.c
+	if ((signo == SIGINT) || (signo == SIGTERM)) {
+		printf("Caught signal %d\n", signo);
+		close(fd);
+		close(sockfd);
+		remove("/var/tmp/aesdsocketdata");
+		delete_all_the_memory();
+		exit (0);
```

Comment:
@srki3050 the `exit` here bypasses your other cleanup path, and explains the memory leak valgrind is showing you.  You won't be guaranteed to free the memory allocated for your thread if the signal occurs between the allocation of the data and the addition to your thread linked list.